Wobble/Scheduling/TaskHandler.cs

```diff
@@ -0,0 +1,119 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Wobble.Logging;
+
+namespace Quaver.Shared.Scheduling
+{
+    /// <summary>
+    ///     This will be used to handle threaded tasks.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TResult"></typeparam>
+    public class TaskHandler<T, TResult> : IDisposable
+    {
+        private Func<T, TResult> Function { get; }
+
+        /// <summary>
+        ///     Event Invoked when the task gets completed.
+        /// </summary>
+        public event EventHandler<TaskCompleteEventArgs<T, TResult>> OnCompleted;
+
+        /// <summary>
+        ///     Event Invoked when the task gets cancelled.
+        /// </summary>
+        public event EventHandler<TaskCancelledEventArgs<T>> OnCancelled;
+
+        /// <summary>
+        ///     Used to handle Task Cancellation.
+        /// </summary>
+        public CancellationTokenSource Source { get; private set; } = new CancellationTokenSource();
+
+        /// <summary>
+        ///     Determined by if current task is completed.
+        /// </summary>
+        public bool IsCompleted { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is running.
+        /// </summary>
+        public bool IsRunning { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is cancelled.
+        /// </summary>
+        public bool IsCancelled { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="function"></param>
+        public TaskHandler(Func<T, TResult> function) => Function = function;
+
+        /// <summary>
+        ///     This method will cancel the previous task and run a new task.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="delay"></param>
+        public void Run(T input, int delay = 0)
+        {
+            // We want to automatically cancel the previous task and dispose the other
+            // cancellation token source before running a new task.
+            Source.Cancel();
+            Source.Dispose();
+            Source = new CancellationTokenSource();
+            IsCompleted = false;
+            IsCancelled = false;
+            IsRunning = true;
+
+            Task.Run(async () =>
+            {
+                try
+                {
+                    var token = Source.Token;
```

Comment:
Retrieve the token before `Task.Run()`. Otherwise there's a race condition: when `Run()` is called a second time before the thread is scheduled, `Source` will get renewed and the `token` will be the new `Source`'s token in both threads, even though the first thread should have got the old `Source`'s token.